lib/hyrax/transactions/transaction.rb

```diff
@@ -0,0 +1,141 @@
+# frozen_string_literal: true
+
+require 'dry/monads'
+require 'dry/monads/do'
+
+module Hyrax
+  module Transactions
+    ##
+    # Queues up a set of `#steps` to run over a value via `#call`. Steps are
+    # given by name, and resolved at runtime via `container`, which defaults to
+    # `Hyrax::Transactions::Container`. This allows injection of specific
+    # handling for named steps at any time.
+    #
+    # `#call` will **ALWAYS** return a `Result` (`Success`|`Failure`) and it's
+    # recommended for users to handle the output in a way that directly
+    # addresses the `Failure` case. The simplest way to do this is to use
+    # `#value_or`: `tx.call(my_value).value_or { |failure| handle_failure(f) }`.
+    #
+    # @since 3.0.0
+    #
+    # @example running a transaction for a set of steps
+    #   steps = ['change_set.validate', 'change_set.save']
+    #   tx    = Hyrax::Transactions::Transaction.new(steps: steps)
+    #
+    #   change_set = Hyrax::ChangeSet.for(Hyrax::Work.new)
+    #   change_set.title = ['comet in moominland']
+    #
+    #   tx.call(change_set) # => Success(#<Hyrax::Work ...>)
+    #
+    # @example with a failure
+    #   class ChangeSetWithTitleValidation < Hyrax::ChangeSet
+    #     self.fields = [:title]
+    #     validates :title, presence: true
+    #   end
+    #
+    #   change_set = ChangeSetWithTitleValidation.new(Hyrax::Work.new)
+    #   change_set.title = []
+    #
+    #   tx.call(change_set)
+    #   # => Failure([:failed_validation, #<Reform::Form::ActiveModel::Errors ...>]
+    #
+    # @example unwapping values safely with handling for failures
+    #   tx.call(change_set).value_or { |failure| "uh oh: #{failure} }
+    #
+    # @example a pattern for subclassing to create new transactions
+    #   class CustomTransaction < Transaction
+    #     DEFAULT_STEPS = ['step.1', 'step.2', 'step.3'].freeze
+    #
+    #     def initialize(container: Container, steps: DEFAULT_STEPS)
+    #       super
+    #     end
+    #   end
+    #
+    #   tx = CustomTransaction.new
+    #   tx.call(:some_value)
+    #
+    class Transaction
+      include Dry::Monads[:result]
+      include Dry::Monads::Do.for(:call)
+
+      ##
+      # @!attribute [rw] container
+      #   @return [Container]
+      # @!attribute [rw] steps
+      #   @return [Array<String>]
+      attr_accessor :container, :steps
+
+      ##
+      # @api public
+      #
+      # @param [Container] container
```

Comment:
Is the Container namespaced?

Reply:
this is a little fussy. `Container` needs to extend `Dry::Container::Mixin`, but it's not actually a class and there's not a common inherited ancestor. The best I could figure was to allow this to resolve to `Hyrax::Transaction::Container` (certainly it should register all the names registered there) but leave it a bit vague since I think it wouldn't be out-of-the-question to slot in a new container matching that interface.

I guess in short, it's really `Hyrax::Transaction::Container` that defines the interface here.